sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -376,6 +387,23 @@ object ScalaReflection extends ScalaReflection {
           dataType = ObjectType(udt.getClass))
         Invoke(obj, "deserialize", ObjectType(udt.userClass), getPath :: Nil)
 
+      case t if isValueClass(t) =>
+        // nested value class is treated as its underlying type
+        // top level value class must be treated as a product
+        val underlyingType = getUnderlyingTypeOf(t)
+        val underlyingClsName = getClassNameFromType(underlyingType)
+        val clsName = t.typeSymbol.asClass.fullName
+        val newTypePath = s"""- Scala value class: $clsName($underlyingClsName)""" +:
+          walkedTypePath
+
+        val arg = deserializerFor(underlyingType, path, newTypePath)
+        if (path.isDefined) {
+          arg
```

Comment:
why can we skip the `NewInstance`?

Reply:
Take class `User` above for example. After compile, field id of type `Id` will become `Int` so when constructing `User` we need `id` to be `Int`.

Also why we need `NewInstance` in case `Id` is itself the schema? Because `Id` may remain as `Id` if it is treated as another type (following [allocation rule](https://docs.scala-lang.org/overviews/core/value-classes.html#allocation-details)). For example, in method [encodeDecodeTest](https://github.com/apache/spark/blob/a40806d2bd84e9a0308165f0d6c97e9cf00aa4a3/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala#L373), if we pass an instance of `Id` as input, it will not be converted to `Int`. In the other case when the required type is explicitly `Id`, then both the input and the result returned from deserialization will both become `Int`.

Reply:
I have the same question. When we deserialize a value class, don't we always have `NewInstance` to construct it back? When constructing `User`, can't we pass in a `Id`?

Reply:
@viirya we cannot because after compile, field `id: Id` becomes `id: Int`

Reply:
So after Scala compile, value class will be its underlying type?

Reply:
yes, but there are some cases where an instance of value class must be created. You can see this [allocation rule](https://docs.scala-lang.org/overviews/core/value-classes.html#allocation-details).

Reply:
> For example, in method encodeDecodeTest, if we pass an instance of Id as input, it will not be converted to Int.

Why in this case, `Id` won't be a Int? It seems to not be in the three rules.

Reply:
it is the 1st rule where `Id` is treated as a type parameter, you can see the `def identity[T]` example

Reply:
hmm, I think the first rule is that a value class is treated as another type. Isn't `T` just `Id`?

Reply:
Oh, sorry I didn't see the identify case.

Reply:
If the `T` is `User`, will `Id` class be instantiated?

Reply:
No in that case `Id` will be `Int`

Reply:
So for Dataset, can we conclude that if a value class is top-level, it will be instantiated, but if it is nested field, it will be just underlying type?

Reply:
yes @viirya